### PR TITLE
Refactor custom_preprocess() for more robustness

### PR DESCRIPTION
**Pull request by dkrako**
Info about PR:

Instead of passing the complete tuple with `model, x_batch, y_batch, a_batch, s_batch, ..` we instead pass a dictionary with the variable names as keys.

If a key matches a variable name like `x_batch`, it will be overwritten.

If a key refers to a new variable, this variable will be passed instance-wise to `evaluate_instance()`.

If we pass None instead of a dictionary in `custom_preprocess`, no changes or additional variables will be passed to evaluate_instance.

Minimum acceptance criteria
- All tests pass

### CODE REVIEW

File: quantus/metrics/axiomatic/completeness.py [hunk 1]
```diff
@@ -255,8 +255,6 @@ def evaluate_instance(
         y: np.ndarray,
         a: np.ndarray,
         s: np.ndarray,
-        c: Any,
```
Comment (annahedstroem) [inline quantus/metrics/axiomatic/completeness.py:258]: Remove from docstring:

c: any
    The custom input to be evaluated on an instance-basis.
p: any
    The custom preprocess input to be evaluated on an instance-basis.

Comment (annahedstroem) [reply quantus/metrics/axiomatic/completeness.py:258]: Replace in all places in docstring so that we don't miss

File: quantus/metrics/base.py [hunk 2]
```diff
@@ -218,52 +212,14 @@ def __call__(
...
-                x=x_instance,
-                y=y_instance,
-                a=a_instance,
-                s=s_instance,
-                c=c_instance,
-                p=p_instance,
-            )
+        iterator = self.get_instance_iterator(data=data)
+        for id_instance, data_instance in iterator:
+            result = self.evaluate_instance(i=id_instance, **data_instance)
```
Comment (annahedstroem) [inline quantus/metrics/base.py:218]: So nice.

File: quantus/metrics/base.py [hunk 3]
```diff
@@ -218,52 +212,14 @@ def __call__(
...
-                i=int(id_instance),
-                model=model,
-                x=x_instance,
-                y=y_instance,
-                a=a_instance,
-                s=s_instance,
-                c=c_instance,
-                p=p_instance,
-            )
+        iterator = self.get_instance_iterator(data=data)
```
Comment (annahedstroem) [inline quantus/metrics/base.py:216]: We need to write a docsting a la numpydoc for this function

File: quantus/metrics/base.py [hunk 4]
```diff
@@ -218,52 +212,14 @@ def __call__(
...
         # Call custom post-processing.
-        self.custom_postprocess(
-            model=model,
-            x_batch=x_batch,
-            y_batch=y_batch,
-            a_batch=a_batch,
-            s_batch=s_batch,
-            custom_batch=custom_batch,
-        )
+        self.custom_postprocess(**data)
```
Comment (annahedstroem) [inline quantus/metrics/base.py:222]: :)

File: quantus/metrics/base.py [hunk 5]
```diff
@@ -287,14 +243,12 @@ def __call__(
     @abstractmethod
     def evaluate_instance(
         self,
-        i: int,
+        i: int,  # TODO: remove this from the general case and check why we need this workaround.
```
Comment (annahedstroem) [inline quantus/metrics/base.py:246]: Not sure what is meant by the TODO comment?

**Review by annahedstroem — COMMENTED** [review]
Please see the comments. Some require changes on several places, like update docstrings of Returns and Params etc. Thank you!

File: quantus/metrics/axiomatic/completeness.py (hunk 1 above)
Comment (dkrako) [reply quantus/metrics/axiomatic/completeness.py:258]: done

File: quantus/metrics/base.py (hunk 2 above)
Comment (dkrako) [reply quantus/metrics/base.py]: :)

File: quantus/metrics/base.py (hunk 3 above)
Comment (dkrako) [reply quantus/metrics/base.py:216]: done

File: quantus/metrics/base.py (hunk 4 above)
Comment (dkrako) [reply quantus/metrics/base.py:222]: thx so much

File: quantus/metrics/base.py (hunk 5 above)
Comment (dkrako) [reply quantus/metrics/base.py]: i have now removed this argument from the default arguments, as there are only two metrics which use this.
we have this new `custom_preprocess()` functionality and I used it for adding `i_batch` and the rest of the attributes.

File: quantus/metrics/localisation/top_k_intersection.py [hunk 6]
```diff
@@ -315,20 +307,8 @@ def custom_preprocess(
             In addition to the x_batch, y_batch, a_batch, s_batch and custom_batch,
```
[the remark this answers is not in the document]
Comment (dkrako) [reply quantus/metrics/localisation/top_k_intersection.py]: done

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -139,7 +139,7 @@ The following methods are expected to be implemented in the metric class:
 - `evaluate_instance()`: Gets model and data for a single instance as input, returns evaluation result.
 
 The following methods are optimal for implementation:
-- `custom_preprocess()`: In case `general_preprocess()` from base class is not sufficient, additional preprocessing steps can be added here.
+- `custom_preprocess()`: In case `general_preprocess()` from base class is not sufficient, additional preprocessing steps can be added here. This method must return a dictionary with string keys or None. If a dictionary is returned, additional keyword arguments can be used in `evaluate_instance()`.
```
[the remark this answers is not in the document]
Comment (dkrako) [reply CONTRIBUTING.md]: done

**Comment by dkrako** [comment]
> # [Codecov](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab) Report
> 
> > Merging [#167](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab) ([9bf05fe](https://github.com/understandable-machine-intelligence-lab/Quantus/commit/9bf05fef1558cacc4607b78a87a6cb428ef872c2)) into [main](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/commit/eec1eef27f8934c30fb3c22a22195e6a97fd1cb6?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab) ([eec1eef](https://github.com/understandable-machine-intelligence-lab/Quantus/commit/eec1eef27f8934c30fb3c22a22195e6a97fd1cb6)) will **decrease** coverage by `0.11%`.
> > The diff coverage is `100.00%`.
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main     #167      +/-   ##
> ==========================================
> - Coverage   94.62%   94.51%   -0.12%     
> ==========================================
>   Files          51       51              
>   Lines        2420     2370      -50     
> ==========================================
> - Hits         2290     2240      -50     
>   Misses        130      130              
> ```
> 
> [Impacted Files](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab) 	Coverage Δ 	
> [quantus/metrics/axiomatic/completeness.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2F4aW9tYXRpYy9jb21wbGV0ZW5lc3MucHk=) 	`94.28% <ø> (ø)` 	
> [quantus/metrics/axiomatic/input_invariance.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2F4aW9tYXRpYy9pbnB1dF9pbnZhcmlhbmNlLnB5) 	`100.00% <ø> (ø)` 	
> [quantus/metrics/axiomatic/non_sensitivity.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2F4aW9tYXRpYy9ub25fc2Vuc2l0aXZpdHkucHk=) 	`100.00% <ø> (ø)` 	
> [quantus/metrics/complexity/complexity.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2NvbXBsZXhpdHkvY29tcGxleGl0eS5weQ==) 	`100.00% <ø> (ø)` 	
> [quantus/metrics/complexity/effective_complexity.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2NvbXBsZXhpdHkvZWZmZWN0aXZlX2NvbXBsZXhpdHkucHk=) 	`100.00% <ø> (ø)` 	
> [quantus/metrics/complexity/sparseness.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2NvbXBsZXhpdHkvc3BhcnNlbmVzcy5weQ==) 	`100.00% <ø> (ø)` 	
> [...s/metrics/faithfulness/faithfulness_correlation.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2ZhaXRoZnVsbmVzcy9mYWl0aGZ1bG5lc3NfY29ycmVsYXRpb24ucHk=) 	`95.91% <ø> (-0.17%)` 	arrow_down
> [...ntus/metrics/faithfulness/faithfulness_estimate.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2ZhaXRoZnVsbmVzcy9mYWl0aGZ1bG5lc3NfZXN0aW1hdGUucHk=) 	`98.00% <ø> (-0.08%)` 	arrow_down
> [quantus/metrics/faithfulness/infidelity.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2ZhaXRoZnVsbmVzcy9pbmZpZGVsaXR5LnB5) 	`96.87% <ø> (-0.10%)` 	arrow_down
> [quantus/metrics/faithfulness/irof.py](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab#diff-cXVhbnR1cy9tZXRyaWNzL2ZhaXRoZnVsbmVzcy9pcm9mLnB5) 	`98.07% <ø> (-0.08%)` 	arrow_down
> ... and [23 more](https://codecov.io/gh/understandable-machine-intelligence-lab/Quantus/pull/167/diff?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab) 		
> 
> Help us with your feedback. Take ten seconds to tell us [how you rate us](https://about.codecov.io/nps?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab). Have a feature suggestion? [Share it here.](https://app.codecov.io/gh/feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=understandable-machine-intelligence-lab)

Coverage seems to have only decreased because this PR has a lot of deletions, so the absolute number of hits has decreased.
The absolute number of misses has not increased though, just the overall percentage of misses in relation to hits.

File: quantus/metrics/base.py (hunk 5 above)
Comment (annahedstroem) [reply quantus/metrics/base.py]: Agree on that solution!

**Review by annahedstroem — COMMENTED** [review]
Great, let's merge!

**Pull request merged**